logstash-core/src/main/java/org/logstash/execution/ShutdownWatcherExt.java

```diff
@@ -105,29 +104,35 @@ public IRubyObject shutdownStalled(final ThreadContext context) {
         if (reports.size() != reportEvery) {
             return context.fals;
         }
-        final int[] inflightCounts = reports.stream().mapToInt(
-            obj -> obj.callMethod(context, "inflight_count").convertToInteger().getIntValue()
-        ).toArray();
-        boolean stalled = true;
+
+        final int[] inflightCounts = reports.stream()
+                .mapToInt(obj -> obj.callMethod(context, "inflight_count").convertToInteger().getIntValue())
+                .toArray();
+
         for (int i = 0; i < inflightCounts.length - 1; ++i) {
             if (inflightCounts[i] > inflightCounts[i + 1]) {
-                stalled = false;
-                break;
+                return context.fals;
             }
         }
-        if (stalled) {
-            final IRubyObject[] stallingThreads = reports.stream().map(
-                obj -> obj.callMethod(context, "stalling_threads")
-            ).toArray(IRubyObject[]::new);
-            for (int i = 0; i < stallingThreads.length - 1; ++i) {
-                if (!stallingThreads[i].op_equal(context, stallingThreads[i + 1]).isTrue()) {
-                    stalled = false;
-                    break;
-                }
+
+        final IRubyObject[] stallingThreads = reports.stream()
+                .map(obj -> obj.callMethod(context, "stalling_threads"))
+                .toArray(IRubyObject[]::new);
+
+        if (inflightCounts.length == 0 && stallingThreads.length == 0) {
+            return context.fals;
+        }
+
+        for (int i = 0; i < stallingThreads.length - 1; ++i) {
+            if (!stallingThreads[i].op_equal(context, stallingThreads[i + 1]).isTrue()) {
+                return context.fals;
             }
-            return stalled ? context.tru : context.fals;
         }
-        return context.fals;
+
+        final int lastInflightCount = inflightCounts.length > 0 ? inflightCounts[inflightCounts.length - 1] : 0;
+        return lastInflightCount > 0 || stallingThreads.length > 0
+                ? context.tru
+                : context.fals;
```

Comment:
if all inflight counts are 0, then we look at `stallingthreads`, but that will always have a length of 5, so it's always true. We'd have to look at the last element of that array instead